File: typescript/api-ts/src/moderation.ts
import OpenAI from "openai";
import { enableModeration } from "./config.js";

type ModerationResult = { flagged: false } | { flagged: true; reason: string };

let _openai: OpenAI | null = null;

function getClient(): OpenAI | null {
  if (!process.env.OPENAI_API_KEY) return null;
  if (!_openai) {
    _openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
  }
  return _openai;
}

/**
 * Check texts against OpenAI's Moderation API.
 * Only runs when ENABLE_MODERATION=true and OPENAI_API_KEY is set.
 */
export async function moderateContent(texts: string[]): Promise<ModerationResult> {
  if (!enableModeration) return { flagged: false };

  const client = getClient();
  if (!client || texts.length === 0) return { flagged: false };

  try {
    const response = await client.moderations.create({ input: texts });
    const hit = response.results.find((r) => r.flagged);
    if (!hit) return { flagged: false };

    const categories = Object.entries(hit.categories)
      .filter(([, v]) => v)
      .map(([k]) => k.replace("/", " / "));

    return {
      flagged: true,
      reason: `Your prompt was flagged for inappropriate content (${categories.join(", ")}). Please revise and try again.`,
    };
  } catch (err) {
    console.error("[moderation] OpenAI moderation check failed:", err);
    return { flagged: false };
  }
}

File: python/api-python/moderation.py
"""Content moderation via OpenAI's Moderation API."""

import os

import httpx
from config import ENABLE_MODERATION

OPENAI_API_KEY = os.getenv("OPENAI_API_KEY", "")


async def moderate_content(texts: list[str]) -> tuple[bool, str | None]:
    """Check texts against OpenAI's Moderation API.

    Only runs when ENABLE_MODERATION=true and OPENAI_API_KEY is set.
    Returns (flagged, reason) tuple.
    """
    if not ENABLE_MODERATION or not OPENAI_API_KEY or not texts:
        return False, None

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            resp = await client.post(
                "https://api.openai.com/v1/moderations",
                headers={"Authorization": f"Bearer {OPENAI_API_KEY}"},
                json={"input": texts},
            )
            resp.raise_for_status()
            data = resp.json()

        for result in data.get("results", []):
            if result.get("flagged"):
                cats = [
                    k.replace("/", " / ")
                    for k, v in result.get("categories", {}).items()
                    if v
                ]
                reason = (
                    f"Your prompt was flagged for inappropriate content"
                    f" ({', '.join(cats)}). Please revise and try again."
                )
                return True, reason

        return False, None
    except Exception as exc:
        print(f"[moderation] OpenAI moderation check failed: {exc}")
        return False, None
